### PR TITLE
[clang][ASTImporter] Improve structural equivalence of overloadable operators.

### PR DESCRIPTION
**Pull request by balazske**
Operators that are overloadable may be parsed as `CXXOperatorCallExpr` or as `UnaryOperator` (or `BinaryOperator`). This depends on the context and can be different if a similar construct is imported into an existing AST. The two "forms" of the operator call AST nodes should be detected as equivalent to allow AST import of these cases.

This fix has probably other consequences because if a structure is imported that has `CXXOperatorCallExpr` into an AST with an existing similar structure that has `UnaryOperator` (or binary), the additional data in the `CXXOperatorCallExpr` node is lost at the import (because the existing node will be used). I am not sure if this can cause problems.

### CODE REVIEW

File: clang/unittests/AST/StructuralEquivalenceTest.cpp [hunk 1]
```diff
@@ -2252,6 +2252,176 @@ TEST_F(StructuralEquivalenceStmtTest, UnaryOperatorDifferentOps) {
...
+      class A;
+      template <typename T, T x, T y>
+      void foo(
+        A<T, x + y>
+      );
+      )",
+      R"(
+      struct Bar;
+
+      Bar& operator+(Bar&, Bar&);
```
Comment (DonatNagyE) [inline clang/unittests/AST/StructuralEquivalenceTest.cpp:2393]: Why is the presence of this `operator+` relevant if this TU doesn't use `operator+` at all?

File: clang/lib/AST/ASTStructuralEquivalence.cpp [hunk 2]
```diff
@@ -437,12 +439,67 @@ class StmtComparer {
...
 /// Determine structural equivalence of two statements.
 static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      const Stmt *S1, const Stmt *S2) {
   if (!S1 || !S2)
     return S1 == S2;
 
+  // Check for statements with similar syntax but different AST.
+  // The unary and binary operators (like '+', '&') can be parsed as
+  // CXXOperatorCall too (and UnaryOperator or BinaryOperator).
+  // This depends on arguments with unresolved type and on the name lookup.
```
Comment (DonatNagyE) [inline clang/lib/AST/ASTStructuralEquivalence.cpp:485]: Could you perhaps elaborate this sentence a bit? Do I understand it correctly that e.g. in `CXXOperatorCallExprVsUnaryBinaryOperator` the presence of `struct Bar` in the second TU ensured that all the operators are parsed as `CXXOperatorCall`s, while in the first TU they are parsed as unary/binary operator expressions?

**Review by NagyDonat — COMMENTED** [review]
As you write that unifying these different but equivalent expressions could've unintended consequences, perhaps you should check the effects of this change on CTU analysis of modern (a.k.a. crazy) C++ projects (or publish the results if you already did so).

Otherwise I can't find any problem in this code, but I'm not familiar enough with these intricate details of the standard, so this opinion is not too relevant.

File: clang/lib/AST/ASTStructuralEquivalence.cpp (hunk 2 above)
Comment (balazske) [reply clang/lib/AST/ASTStructuralEquivalence.cpp:485]: The idea was coming from this post:
https://discourse.llvm.org/t/different-ast-translations-for-the-same-body-why/22741
The presence of `operator+` declaration is needed to change the parsed AST from `BinaryOperator` to `CXXOverloadedOperatorCallExpr`.

File: clang/unittests/AST/StructuralEquivalenceTest.cpp (hunk 1 above)
Comment (balazske) [reply clang/unittests/AST/StructuralEquivalenceTest.cpp:2393]: I do not know exactly what was the purpose, but likely it should be `operator -` instead.

File: clang/lib/AST/ASTStructuralEquivalence.cpp (hunk 2 above)
Comment (DonatNagyE) [reply clang/lib/AST/ASTStructuralEquivalence.cpp:485]: Thanks for the link, it's really helpful. Perhaps it should be included in the source code comment as well, to help those who'll work on this code in the future.

Comment (DonatNagyE) [reply clang/lib/AST/ASTStructuralEquivalence.cpp]: Thanks, the updated comment is clear now.

**Review by NagyDonat — APPROVED** [review]
Thanks for the clarifications, it looks good to me now. (Except for the `FIXME`, but that's an independent issue and you're right that it should be handled in another commit.)

**Pull request merged**